File: src/ai-train/predict_move_quality.py
import sys
import numpy as np
import chess
import tensorflow as tf

# Load mô hình đã huấn luyện
model = tf.keras.models.load_model('model_move_quality.h5')

def fen_to_matrix(fen):
    board = chess.Board(fen)
    piece_map = board.piece_map()
    matrix = np.zeros((8, 8), dtype=int)
    for square, piece in piece_map.items():
        row = 7 - (square // 8)
        col = square % 8
        matrix[row, col] = piece.piece_type * (1 if piece.color else -1)
    return matrix.flatten()

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("0.0")
        sys.exit(1)
    fen = sys.argv[1]
    x = np.array([fen_to_matrix(fen)])
    score = model.predict(x)[0][0]
    print(f"{score}")

File: src/ai-train/predict_move_quality.ts
import * as tf from "@tensorflow/tfjs-node";
import { Chess } from "chess.js";

// Load model (async)
export async function predictMoveQuality(fen: string): Promise<number> {
  const model = await tf.loadLayersModel("file://app/model/model.json");

  function fenToMatrix(fen: string): number[] {
    const board = new Chess(fen);
    const pieceMap = board.board();
    const matrix: number[] = [];
    for (let row = 7; row >= 0; row--) {
      for (let col = 0; col < 8; col++) {
        const piece = pieceMap[row][col];
        if (!piece) {
          matrix.push(0);
        } else {
          const typeMap: any = { p: 1, n: 2, b: 3, r: 4, q: 5, k: 6 };
          const value =
            typeMap[piece.type.toLowerCase()] * (piece.color === "w" ? 1 : -1);
          matrix.push(value);
        }
      }
    }
    return matrix;
  }

  const input = tf.tensor2d([fenToMatrix(fen)], [1, 64]);
  const prediction = model.predict(input) as tf.Tensor;
  const score = (await prediction.data())[0];
  return score;
}

// Sử dụng ví dụ
const fen = "your_fen_string_here";
predictMoveQuality(fen).then((score) => {
  console.log(score);
});
